www/js/lib/uiUtil.js

```diff
@@ -227,12 +228,92 @@ define([], function() {
+        return string;
+    }
 
+    /**
+     * Applies the requested app and content theme
+     * 
+     * A <theme> string consists of two parts, the appTheme (theme to apply to the app shell only), and an optional
+     * contentTheme beginning with an underscore: e.g. 'dark_invert' = 'dark' (appTheme) + '_invert' (contentTheme)
+     * Current themes are: light, dark, dark_invert, dark_mwInvert but code below is written for extensibility
+     * For each appTheme (except the default 'light'), a corresponding set of rules must be present in app.css
+     * For each contentTheme, a stylesheet must be provided in www/css that is named 'kiwixJS' + contentTheme
+     * A rule may additionally be needed in app.css for full implementation of contentTheme
+     * 
+     * @param {String} theme The theme to apply (light|dark[_invert|_mwInvert])
+     */
+    function applyAppTheme(theme) {
+        var htmlEl = document.getElementsByTagName('html')[0];
+        var oldTheme = htmlEl.dataset.theme || '';
+        var iframe = document.getElementById('articleContent');
+        var doc = iframe.contentDocument;
+        var kiwixJSSheet = doc ? doc.getElementById('kiwixJSTheme') || null : null;
+        var appTheme = theme.replace(/_.*$/, '');
+        var contentTheme = theme.replace(/^[^_]*/, '');
+        var oldAppTheme = oldTheme.replace(/_.*$/, '');
+        var oldContentTheme = oldTheme.replace(/^[^_]*/, '');
+        // Remove oldAppTheme if necessary
+        if (oldAppTheme && oldAppTheme !== appTheme) htmlEl.classList.remove(oldAppTheme);
+        // Apply new appTheme (NB it will not be added twice if it's already there)
+        if (appTheme) htmlEl.classList.add(appTheme);
+        // Embed a reference to applied theme, so we can remove it generically in the future
+        htmlEl.dataset.theme = theme;
+        // Hide any previously displayed help
+        var oldHelp = document.getElementById(oldTheme + '-help');
+        if (oldHelp) oldHelp.style.display = 'none';
+        // Show any specific help for selected contentTheme
+        var help = document.getElementById(theme + '-help');
+        if (help) help.style.display = 'block';
+        
+        // If there is no ContentTheme or we are applying a different ContentTheme, remove any previously applied ContentTheme
+        if (oldContentTheme && oldContentTheme !== contentTheme) {
+            iframe.classList.remove(oldContentTheme);
+            if (kiwixJSSheet) {
+                kiwixJSSheet.disabled = true;
+                kiwixJSSheet.parentNode.removeChild(kiwixJSSheet);
+            }
+        }
+        // Apply the requested ContentTheme (if not already attached)
+        if (contentTheme && (!kiwixJSSheet || !~kiwixJSSheet.href.search(contentTheme))) {
+            iframe.classList.add(contentTheme);
+            // Use an absolute reference becuase Service Worker needs this
+            var prefix = document.location.href.replace(/index\.html(?:$|[#?].*$)/, '');
```

Comment:
@mossroy The issue with SW mode is the usual one concerning ZIMs with subdirectories. If we are in a subdirectory (or even a document that has '/' in its filename, now treated by mwoffliner as a subdirectory), relative links to a stylesheet in the filesystem added after a document has loaded won't work. Since the browser can give us an absolute URL for index.html (NB this is index.html in our filesystem), I thought it best to use that. I believe the URL for the app will always either be `…...……/www/`, or `...………./www/index.html[?#….]`. Unless we ourselves rename index.html to something else.
I suppose there is the risk another dev may rename index.html and then this will break. So I'll check out your alternatives. Thanks for those! (I did consider using `uiUtil.deriveZimUrlFromRelativeUrl()`, but it is coded to work only with ZIM URLs.)

Reply:
`document.location.origin + document.location.pathname` works perfectly.